README.md
# Learning golang

The purpose of this application is to become familiar with the Go language.

The app is called "myapp" and it will surely be an e-commerce application with an API server, a frontend and a mongodb backend

A second objective will be to also provide the k8s deployments in a separated repository (best practice)

Documentation generation and test execution is mandatory

## Requirements

* Golang
* $GOPATH in the $PATH
* Docker-cli

## Build

Some docker images `myapp` are built by the following command:
```shell
make build
```

## Run

The docker image built can be run:
```shell
make run
```

## Help

In order to find the possible targets of the Makefile
```shell
make help
```

CHANGELOG.md
# CHANGELOG

## v0.0.2

### myapp

- FEATURE(api-server): Return a version based on tags and commits
- FEATURE(github-action): Release on tags like prepare-vX.Y.Z



api-server_test.go
package main

import (
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/gin-gonic/gin"
)

func TestListProducts(t *testing.T) {
	// Switch to test mode
	gin.SetMode(gin.TestMode)

	// Setup your router, like  did in the main function
	router := gin.Default()
	v1 := router.Group("/api/v1")
	{
		products := v1.Group("/products")
		{
			products.GET("", ListProducts)
		}
	}

	// Test the route as defined in api-server.go
	req, err := http.NewRequest(http.MethodGet, "/api/v1/products", nil)
	if err != nil {
		t.Fatalf("Couldn't create request: %v\n", err)
	}

	// Create a response recorder to inspect the response
	rec := httptest.NewRecorder()

	// Perform the request
	router.ServeHTTP(rec, req)
	// fmt.Println(rec.Body)

	// Check to see if the response was what you expected
	if rec.Code == http.StatusOK {
		t.Logf("Expected to get status %d is same as %d\n", http.StatusOK, rec.Code)
	} else {
		t.Fatalf("Expected to get status %d but instead got %d\n", http.StatusOK, rec.Code)
	}
}

func TestShowProduct(t *testing.T) {
	gin.SetMode(gin.TestMode)
	router := gin.Default()
	v1 := router.Group("/api/v1")
	{
		products := v1.Group("/products")
		{
			products.GET(":id", ShowProduct)
		}
	}

	req, err := http.NewRequest(http.MethodGet, "/api/v1/products/1", nil)
	if err != nil {
		t.Fatalf("Couldn't create request: %v\n", err)
	}
	rec := httptest.NewRecorder()
	router.ServeHTTP(rec, req)
	// test
	if rec.Code == http.StatusOK {
		t.Logf("Expected to get status %d is same as %d\n", http.StatusOK, rec.Code)
	} else {
		t.Fatalf("Expected to get status %d but instead got %d\n", http.StatusOK, rec.Code)
	}
}

func TestPing(t *testing.T) {
	// Switch to test mode
	gin.SetMode(gin.TestMode)

	// Setup your router, like  did in the main function
	router := gin.Default()
	v1 := router.Group("/api/v1")
	{
		v1.GET("/ping", Ping)
	}

	// Test the route as defined in api-server.go
	req, err := http.NewRequest(http.MethodGet, "/api/v1/ping", nil)
	if err != nil {
		t.Fatalf("Couldn't create request: %v\n", err)
	}

	// Create a response recorder to inspect the response
	rec := httptest.NewRecorder()

	// Perform the request
	router.ServeHTTP(rec, req)

	// Check to see if the response was what you expected
	if rec.Code == http.StatusOK {
		t.Logf("Expected to get status %d is same as %d\n", http.StatusOK, rec.Code)
	} else {
		t.Fatalf("Expected to get status %d but instead got %d\n", http.StatusOK, rec.Code)
	}
}

func TestVersion(t *testing.T) {
	// Switch to test mode
	gin.SetMode(gin.TestMode)

	// Setup your router, like did in the main function
	router := gin.Default()
	v1 := router.Group("/api/v1")
	{
		v1.GET("/version", GetVersion)
	}

	// Test the route as defined in api-server.go
	req, err := http.NewRequest(http.MethodGet, "/api/v1/version", nil)
	if err != nil {
		t.Fatalf("Couldn't create request: %v\n", err)
	}

	// Create a response recorder to inspect the response
	rec := httptest.NewRecorder()

	// Perform the request
	router.ServeHTTP(rec, req)

	// Check to see if the response was what you expected
	if rec.Code == http.StatusOK {
		t.Logf("Expected to get status %d is same as %d\n", http.StatusOK, rec.Code)
	} else {
		t.Fatalf("Expected to get status %d but instead got %d\n", http.StatusOK, rec.Code)
	}
}


main.go
package main

import (
	// Import net/http. This is required for example if using constants such as http.StatusOK.
	"net/http"

	"github.com/gin-gonic/gin"

	_ "github.com/cdelgehier/learning_golang/docs" // This line is necessary for go-swagger
	"github.com/cdelgehier/learning_golang/models"
)

// articles slice to seed article data.
var articles = []models.Article{
	{ID: "1", Brand: "Ferrero", Name: "Kinder Bueno", Price: 3.56},
	{ID: "2", Brand: "Mars, Incorporated", Name: "M&M's", Price: 2.43},
	{ID: "3", Brand: "Haribo", Name: "Rotella", Price: 13.90},
}

// @title           MyApp API
// @version         1.0
// @description     This is the MyApp API server.
// @termsOfService  http://swagger.io/terms/

// @contact.name   DELGEHIER Cedric
// @contact.url    https://twitter.com/JackNemrod

// @license.name  Apache 2.0
// @license.url   http://www.apache.org/licenses/LICENSE-2.0.html

// @host      localhost:8080
// @BasePath  /api/v1

func main() {

	// Default With the Logger and Recovery middleware already attached
	router := gin.Default()

	// Group routes by version
	v1 := router.Group("/api/v1")
	{
		articles := v1.Group("/articles")
		{
			// articles.GET(":id", ShowArticle)
			articles.GET("", ListArticles)
		}
	}

	router.Run(":8080")
}

// ListArticles godoc
// @Summary      List articles
// @Description  Get a list of all articles known
// @Tags         articles
// @Accept       json
// @Produce      json
// @Success      200  {array}   models.Article
// @Router       /api/v1/articles [get]
func ListArticles(c *gin.Context) {
	c.IndentedJSON(http.StatusOK, articles)
}


tools/build.sh
#!/bin/bash
set -e -u -o pipefail

CURRENT_DIR="$(cd "$(dirname "$0")" && pwd -P)"
REPOSITORY_ROOT="$(cd "${CURRENT_DIR}/.." && pwd -P)"

ARCHIVE_FILE="${REPOSITORY_ROOT}/${APP_NAME-myapp}-${TAG_NAME-NO_TAGNAME}.tar.gz"

if [ ! -f "${ARCHIVE_FILE}" ]
then
  echo "* Generating release archive"

  TMP_DIR="$(mktemp -d)"
  ARCHIVE_BASENAME="myapp-${TAG_NAME-NO_TAGNAME}"
  TMP_WORKINGDIR="${TMP_DIR}/${ARCHIVE_BASENAME}"

  rm -f "${ARCHIVE_FILE}"
  cd "${TMP_DIR}" || exit 1
  cp -r "${REPOSITORY_ROOT}" "${TMP_WORKINGDIR}"


  # Exclusions avoid not needed technical files to be present in the final archive
  tar -cz \
    --exclude='NOTES.md' \
    --exclude='.gitignore' \
    --exclude='tools' \
    --exclude='.git' \
    --exclude='.github' \
    --exclude='Makefile' \
    --exclude='RELEASE.md' \
    -f "${ARCHIVE_FILE}" "${ARCHIVE_BASENAME}"

  echo "* Done !"
  echo
  echo
  echo "**** Release file is ready in ${ARCHIVE_FILE} ****"
  echo
  echo
  echo
else
  echo "== File ${ARCHIVE_FILE} found: nothing to build."
fi


tools/release.sh
#!/bin/bash
set -x -e -u -o pipefail

CURRENT_DIR="$(cd "$(dirname "$0")" && pwd -P)"
REPOSITORY_ROOT="$(cd "${CURRENT_DIR}/.." && pwd -P)"

## Check requirements
# Git CLI
: "${1?"ERROR: First argument empty. Please provide a valid tag pointing to the current HEAD as argument. Exiting."}"
command -v git >/dev/null 2>&1 || { echo "ERROR: Cannot find the 'git' command line which is required. Exiting." && exit 1;}
# Hub CLI
: "${GITHUB_REPOSITORY?"ERROR: Variable GITHUB_REPOSITORY is empty. Please provide it to allow the 'hub' CLI to work as expected. Exiting."}"
GITHUB_USER="${GITHUB_REPOSITORY%/*}"
GITHUB_PROJECT="${GITHUB_REPOSITORY#*/}"
: "${GITHUB_TOKEN?"ERROR: Variable GITHUB_TOKEN is empty. Please provide it to allow the 'hub' CLI to work as expected. Exiting."}"
command -v hub >/dev/null 2>&1 || { echo "ERROR: Cannot find the 'hub' command line which is required. Exiting." && exit 1;}

## Check that the tag provided as argument matches the expected pattern
[[ "${1}" =~ prepare-v([0-9]+)\.([0-9]+)\.([0-9]+) ]] || \
  { echo "ERROR: Provided argument(${1}) does not match the expected pattern ('vX.Y.Z' where X, Y and Z are integer digits). Exiting." && exit 1;}
PREPARE_TAG="${1}"
FINALE_TAG="${1:8}"
BRANCH="${FINALE_TAG%%\.*}.x" # Branch name is extracted from the major version
echo "== Provided tag '${PREPARE_TAG}' is valid. Preparing a release with the tag '${FINALE_TAG}' on the branch '${BRANCH}'..."

## Generate changelog between current HEAD and previous tag
PREVIOUS_TAG="$(git describe --tags --abbrev=0 HEAD~1)"
[ -n "${PREVIOUS_TAG}" ] || PREVIOUS_TAG="$(git rev-list --max-parents=0 HEAD)"
: "${PREVIOUS_TAG?"ERROR: Could not determine a reference commit neither a previous tag or the oldest commit. Exiting"}"

CURRENT_COMMIT="$(git rev-parse --verify HEAD)"
: "${CURRENT_COMMIT?"ERROR: Could not retrieve neither a previous tag or the oldest commit. Exiting"}"

echo "== Preparing Changelog from ref. ${PREVIOUS_TAG} to current ref. ${CURRENT_COMMIT}..."

# The list of commit from the history is processed with the following:
# - Removing lines with the string "BUMP" (commit tags)
# - Sorting and removing doublons
# - Replacing double spaces by simple spaces
changelog="$(git log "${PREVIOUS_TAG}..${CURRENT_COMMIT}" --pretty="format:- %s" | cat \
  | grep -v "^- BUMP" \
  | sort | uniq \
  | sed 's/  / /g' \
)"

echo "== Changelog retrieved: preparing release commit by writing and commiting files..."
## Prepend changelog message in the changelog file
CHANGELOG_FILE="${REPOSITORY_ROOT}/CHANGELOG.md"
sed -i '/# CHANGELOG/r'<(printf "\n## %s\n\n### myapp\n\n%s\n" "${FINALE_TAG}" "${changelog}") "${CHANGELOG_FILE}"

## Write expected version in the VERSION file
VERSION_FILE="${REPOSITORY_ROOT}/VERSION"
echo "${FINALE_TAG}" > "${VERSION_FILE}"

## Commit and push changes Changes
git config --local user.name "CI"
git config --local user.email "ci@worker.local"
git add "${CHANGELOG_FILE}" "${VERSION_FILE}"
git commit -m "BUMP: Update VERSION and CHANGELOG.md for ${FINALE_TAG}"

echo "== Release commit ready. Pushing changes to repository and creating release..."
git checkout -b "${BRANCH}"
git push origin "HEAD:${BRANCH}"

## Remove "prepare" tag
git push --delete origin "${PREPARE_TAG}"

## Generate a new archive including release's changes with the finale tag name
# Note that the variable TAG_NAME is overriden to use the finale tag (and not the "prepare-..." tag)
TAG_NAME="${FINALE_TAG}" bash "${CURRENT_DIR}/build.sh"
# Retrieve the name of this new archive with the same "TAG_NAME overriding" technique
ARCHIVE_FILE="${REPOSITORY_ROOT}/${APP_NAME-myapp}-${FINALE_TAG-NO_TAGNAME}.tar.gz"

## Create a Release (along with a tag) in Github from the current commit
RELEASE_COMMIT="$(git rev-parse --verify HEAD)"
printf "%s\n\n## %s\n\n%s\n" "${FINALE_TAG}" "Changes since ${PREVIOUS_TAG}" "${changelog}" \
  | hub release create --attach="${ARCHIVE_FILE}" --file=- --commitish "${RELEASE_COMMIT}" "${FINALE_TAG}"

echo "== Release finished successfully"
exit 0


api-server.go
// Golang REST API server
package main

import (
	// Import net/http. This is required for example if using constants such as http.StatusOK.
	"net/http"
	"time"

	"github.com/gin-gonic/gin"

	_ "github.com/cdelgehier/learning_golang/docs" // This line is necessary for go-swagger
	"github.com/cdelgehier/learning_golang/models"
)

// products slice to seed product data.
var products = []models.Product{
	{ID: "1", Brand: "Ferrero", Name: "Kinder Bueno", Price: 3.56},
	{ID: "2", Brand: "Mars, Incorporated", Name: "M&M's", Price: 2.43},
	{ID: "3", Brand: "Haribo", Name: "Rotella", Price: 13.90},
}

var (
	Version   string
	Build     string
	BuildTime string
)

// @title           MyApp API
// @version         Version
// @description     This is the MyApp API server.
// @termsOfService  http://swagger.io/terms/

// @contact.name   DELGEHIER Cedric
// @contact.url    https://twitter.com/JackNemrod

// @license.name  Apache 2.0
// @license.url   http://www.apache.org/licenses/LICENSE-2.0.html

// @host      localhost:8080
// @BasePath  /api/v1

func main() {

	// Default With the Logger and Recovery middleware already attached
	router := gin.Default()

	// Group routes by version
	v1 := router.Group("/api/v1")
	{
		v1.GET("/ping", Ping)
		v1.GET("/version", GetVersion)

		products := v1.Group("/products")
		{
			products.GET(":id", ShowProduct)
			products.GET("", ListProducts)
		}
	}

	router.Run(":8080")
}

// ListProducts godoc
// @Summary      List products
// @Description  Get a list of all products known
// @Tags         products
// @Accept       json
// @Produce      json
// @Success      200  {array}   models.Product
// @Router       /api/v1/products [get]
func ListProducts(c *gin.Context) {
	c.IndentedJSON(http.StatusOK, products)
}

// ShowProduct godoc
// @Summary      Show product by ID
// @Description  Get a product by its ID
// @Tags         products
// @Param 		 id path string true "The ID of the product"
// @Accept       json
// @Produce      json
// @Success      200  {object}   models.Product
// @Failure 	 404  {string}   Error "error"
// @Router       /api/v1/products/{id} [get]
func ShowProduct(c *gin.Context) {
	id := c.Param("id")

	for _, p := range products {
		if p.ID == id {
			c.IndentedJSON(http.StatusOK, p)
			return
		}
	}
	c.IndentedJSON(http.StatusNotFound, gin.H{"message": "product not found"})

}

// Ping godoc
// @Summary      Ping
// @Description  Ping method for API Server
// @Accept       json
// @Produce      json
// @Success      200  {object} models.Ping "ping !"
// @Router       /api/v1/ping [get]
func Ping(c *gin.Context) {
	c.IndentedJSON(http.StatusOK, models.Ping{Pong: time.Now()})
}

// Version godoc
// @Summary      Version
// @Description  Version of API-server
// @Accept       json
// @Produce      json
// @Success      200  {string} Version "version"
// @Router       /api/v1/version [get]
func GetVersion(c *gin.Context) {
	c.IndentedJSON(
		http.StatusOK,
		gin.H{
			"version":    Version,
			"build":      Build,
			"build_time": BuildTime,
		})
}
